src/geometry/LinearPropagator.i.hh

```diff
@@ -0,0 +1,64 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file LinearPropagation.i.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "geometry/LinearPropagator.hh"
+#include "base/ArrayUtils.hh"
+#include "geometry/Types.hh"
+#include "physics/base/Units.hh"
+
+namespace celeritas
+{
+//___________________________________________________________________________//
+/*!
+ * Do straight propagation to physics process or boundary and reduce next_step
+ *
+ * Scalar geometry length computation. The track is moved along track.dir()
+ * direction by a distance track.next_step()
+ */
+CELER_FORCEINLINE_FUNCTION
+void LinearPropagator::fast_linear_step(real_type step)
+{
+    axpy(step, track_.dir(), &track_.pos());
+    track_.next_step() -= step;
```

Comment:
I think directly exposing and modifying the track state is dangerous here, since (for example) there is no check that the next step remains nonnegative. But since the main utility of substepping will be in the magnetic field propagator we can defer till then.

Reply:
My thought was to keep this function fast.  There are validity checks in `operator()(dist)`, and one other function for default behavior to move to next boundary, which uses internal variable `next_step` and then `move_next_volume()` right away.

Reply:
We will move this into the base class as needed when implementing the magnetic field propagator.